src/annotator/anchoring/trim-range.ts

```diff
@@ -0,0 +1,225 @@
+  }
+
+  const offsetDelta =
+    availableChars.length - availableNonWhitespaceChars.length;
+
+  return direction === TrimDirection.Backwards
+    ? baseOffset - offsetDelta
+    : baseOffset + offsetDelta;
+}
+
+/**
+ * Calculate a new Range start position (TrimDirection.Forwards) or end position
+ * (Backwards) for `range` that represents the nearest non-whitespace character,
+ * moving into the `range` away from the relevant initial boundary node towards
+ * the terminating boundary node.
+ *
+ * The search is non-inclusive of the initial boundary node, but does consider
+ * the terminating boundary node.
+ *
+ * @throws {RangeError} If no text node with non-whitespace characters found
+ */
+function closestNonSpacePosition(
+  range: Range,
+  direction: TrimDirection
+): RangePosition {
+  const nodeIter =
+    range.commonAncestorContainer.ownerDocument!.createNodeIterator(
+      range.commonAncestorContainer,
+      NodeFilter.SHOW_TEXT
+    );
+
+  const initialBoundaryNode =
+    direction === TrimDirection.Forwards
+      ? range.startContainer
+      : range.endContainer;
+
+  const terminalBoundaryNode =
+    direction === TrimDirection.Forwards
+      ? range.endContainer
+      : range.startContainer;
+
+  let currentNode = nodeIter.nextNode();
+
+  // Advance the NodeIterator to the `initialBoundaryNode`
+  while (currentNode && currentNode !== initialBoundaryNode) {
+    currentNode = nodeIter.nextNode();
+  }
+
+  if (direction === TrimDirection.Backwards) {
+    // Reverse the NodeIterator direction. This will return the same node
+    // as the previous `nextNode()` call (initial boundary node).
+    currentNode = nodeIter.previousNode();
+  }
+
+  let trimmedOffset = -1;
+
+  const advance = () => {
+    currentNode =
+      direction === TrimDirection.Forwards
+        ? nodeIter.nextNode()
+        : nodeIter.previousNode();
+
+    if (currentNode) {
+      const nodeText = currentNode.textContent!;
+      const baseOffset =
+        direction === TrimDirection.Forwards ? 0 : nodeText.length;
+      trimmedOffset = closestNonSpaceOffset(nodeText, baseOffset, direction);
+    }
+  };
+
+  // Start the examination at the first text node before/after the
+  // initial boundary node, i.e. do not evaluate the `container` itself
+  advance();
+
+  while (
+    currentNode &&
+    trimmedOffset === -1 &&
+    currentNode !== terminalBoundaryNode
+  ) {
+    advance();
+  }
+
+  if (currentNode && trimmedOffset >= 0) {
+    return { node: currentNode, offset: trimmedOffset };
+  }
+  /* istanbul ignore next */
+  throw new RangeError('No text nodes with non-whitespace text found in range');
+}
+
+/**
+ * Return a new DOM Range that adjusts the start and end positions of `range` as
+ * needed such that:
+ *
+ * - `startContainer` and `endContainer` text nodes both contain at least one
+ *   non-whitespace character within the Range's text content
+ * - `startOffset` and `endOffset` both point at non-whitespace characters
```

Comment:
To clarify, `startOffset` will be just before a non-whitespace character, and `endOffset` will be just after?

Reply:
Yep, and I've updated the comment:

```
 * - `startOffset` and `endOffset` both reference non-whitespace characters,
 *   with `startOffset` immediately before the first non-whitespace character
 *   and `endOffset` immediately after the last
 * ```